File: android/app/src/main/java/com/raban/etabli/probe/net/WebDavClient.kt
package com.raban.etabli.probe.net

import android.util.Base64
import com.raban.etabli.probe.engine.WebDavCreds
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.xmlpull.v1.XmlPullParser
import java.io.ByteArrayInputStream
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.net.URLDecoder
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

// Minimal Nextcloud WebDAV client — pure-Android, HttpURLConnection + the
// stdlib XmlPullParser. Auth = HTTP Basic with the user's Nextcloud APP
// PASSWORD. We never disable TLS verification; if the cert is bad the
// error surfaces verbatim and the user fixes it server-side.

data class DavEntry(
    val href: String,
    val name: String,
    val isDirectory: Boolean,
    val size: Long?,
    val modifiedAt: Long?,
)

sealed class WebDavError(message: String) : RuntimeException(message) {
    object NotConfigured : WebDavError("Set Nextcloud credentials in Settings.")
    class Http(val status: Int, val body: String?) : WebDavError("Server returned HTTP $status.")
    class Transport(msg: String) : WebDavError("Network error: $msg.")
    class Decoding(msg: String) : WebDavError("Couldn't parse response: $msg.")
}

class WebDavClient(private val creds: WebDavCreds) {

    private fun userRoot(path: String = ""): URL {
        val base = creds.baseURL.trimEnd('/', ' ')
        val cleaned = path.trim('/')
        val suffix = if (cleaned.isEmpty()) "" else "/$cleaned"
        return URL("$base/remote.php/dav/files/${creds.username}$suffix")
    }

    private fun basicAuth(): String {
        val raw = "${creds.username}:${creds.appPassword}".toByteArray(Charsets.UTF_8)
        return "Basic " + Base64.encodeToString(raw, Base64.NO_WRAP)
    }

    suspend fun testConnection(): Int = list("").size

    suspend fun list(path: String): List<DavEntry> = withContext(Dispatchers.IO) {
        val url = userRoot(path)
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "PROPFIND"
            doOutput = true
            instanceFollowRedirects = false
            connectTimeout = 15_000
            readTimeout = 30_000
            setRequestProperty("Authorization", basicAuth())
            setRequestProperty("Depth", "1")
            setRequestProperty("Content-Type", "application/xml; charset=utf-8")
        }
        try {
            OutputStreamWriter(conn.outputStream, Charsets.UTF_8).use { w ->
                w.write(propfindBody)
            }
            val code = conn.responseCode
            if (code != 207 && code != 200) {
                throw WebDavError.Http(code, conn.errorStream?.bufferedReader()?.readText())
            }
            val bytes = conn.inputStream.use { it.readBytes() }
            parseMultiStatus(bytes, selfPath = url.path)
        } catch (e: WebDavError) { throw e }
        catch (t: Throwable) { throw WebDavError.Transport(t.message ?: "?") }
        finally { conn.disconnect() }
    }

    suspend fun get(path: String): ByteArray = withContext(Dispatchers.IO) {
        val url = userRoot(path)
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 60_000
            setRequestProperty("Authorization", basicAuth())
        }
        try {
            val code = conn.responseCode
            if (code != 200) {
                throw WebDavError.Http(code, conn.errorStream?.bufferedReader()?.readText())
            }
            conn.inputStream.use { it.readBytes() }
        } catch (e: WebDavError) { throw e }
        catch (t: Throwable) { throw WebDavError.Transport(t.message ?: "?") }
        finally { conn.disconnect() }
    }

    suspend fun put(targetDir: String, fileName: String, bytes: ByteArray,
                    contentType: String = "application/octet-stream") = withContext(Dispatchers.IO) {
        val cleaned = targetDir.trim('/')
        if (cleaned.isNotEmpty()) {
            val parts = cleaned.split('/')
            for (i in 1..parts.size) mkcol(parts.take(i).joinToString("/"))
        }
        val dest = if (cleaned.isEmpty()) fileName else "$cleaned/$fileName"
        val url = userRoot(dest)
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "PUT"
            doOutput = true
            connectTimeout = 15_000
            readTimeout = 60_000
            setFixedLengthStreamingMode(bytes.size)
            setRequestProperty("Authorization", basicAuth())
            setRequestProperty("Content-Type", contentType)
        }
        try {
            conn.outputStream.use { it.write(bytes) }
            val code = conn.responseCode
            if (code !in 200..299) {
                throw WebDavError.Http(code, conn.errorStream?.bufferedReader()?.readText())
            }
        } catch (e: WebDavError) { throw e }
        catch (t: Throwable) { throw WebDavError.Transport(t.message ?: "?") }
        finally { conn.disconnect() }
    }

    private suspend fun mkcol(path: String) = withContext(Dispatchers.IO) {
        val url = userRoot(path)
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "MKCOL"
            connectTimeout = 15_000
            readTimeout = 30_000
            setRequestProperty("Authorization", basicAuth())
        }
        try {
            val code = conn.responseCode
            if (code == 405 || code == 409) return@withContext      // already exists
            if (code !in 200..299) {
                throw WebDavError.Http(code, conn.errorStream?.bufferedReader()?.readText())
            }
        } catch (e: WebDavError) { throw e }
        catch (t: Throwable) { throw WebDavError.Transport(t.message ?: "?") }
        finally { conn.disconnect() }
    }

    private fun parseMultiStatus(xml: ByteArray, selfPath: String): List<DavEntry> {
        val parser = android.util.Xml.newPullParser()
        parser.setInput(ByteArrayInputStream(xml), "UTF-8")
        val httpDate = SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss zzz", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("GMT")
        }
        val entries = mutableListOf<DavEntry>()
        var inResponse = false
        var href: String? = null
        var displayname: String? = null
        var size: Long? = null
        var lastMod: String? = null
        var isDir = false
        var event = parser.eventType
        while (event != XmlPullParser.END_DOCUMENT) {
            when (event) {
                XmlPullParser.START_TAG -> when (parser.name?.lowercase()) {
                    "response" -> { inResponse = true; href = null; displayname = null; size = null; lastMod = null; isDir = false }
                    "href" -> if (inResponse) href = parser.nextText()?.trim()
                    "displayname" -> if (inResponse) displayname = parser.nextText()?.trim()
                    "getcontentlength" -> if (inResponse) size = parser.nextText()?.toLongOrNull()
                    "getlastmodified" -> if (inResponse) lastMod = parser.nextText()?.trim()
                    "collection" -> if (inResponse) isDir = true
                }
                XmlPullParser.END_TAG -> if (parser.name?.lowercase() == "response" && inResponse) {
                    inResponse = false
                    val raw = href
                    if (raw != null) {
                        val pathOnly = try { URI(raw).path ?: raw } catch (_: Throwable) { raw }
                        val self0 = selfPath.trim('/')
                        val cand0 = pathOnly.trim('/')
                        if (cand0 != self0) {
                            val decoded = try { URLDecoder.decode(raw, "UTF-8") } catch (_: Throwable) { raw }
                            val nameTrail = decoded.trimEnd('/').substringAfterLast('/')
                            val name = displayname?.takeIf { it.isNotBlank() } ?: nameTrail
                            val mod = lastMod?.let { runCatching { httpDate.parse(it)?.time }.getOrNull() }
                            entries += DavEntry(decoded, name, isDir, size, mod)
                        }
                    }
                }
            }
            event = parser.next()
        }
        return entries.sortedWith(
            compareByDescending<DavEntry> { it.isDirectory }.thenBy { it.name.lowercase() }
        )
    }

    companion object {
        private val propfindBody = """<?xml version="1.0"?>
            |<d:propfind xmlns:d="DAV:">
            |  <d:prop>
            |    <d:displayname/>
            |    <d:getcontentlength/>
            |    <d:getlastmodified/>
            |    <d:resourcetype/>
            |  </d:prop>
            |</d:propfind>""".trimMargin()
    }
}

File: ios/App/Net/WebDavClient.swift
import Foundation

// Minimal Nextcloud WebDAV client. URLSession only — no third-party deps,
// matches the rest of the suite. We use HTTP Basic with the user's Nextcloud
// APP PASSWORD (not their login). NSURLSession honours System Trust by
// default, so a self-signed cert raises an error — we never override the
// challenge handler to "accept anything" and you should never do that here
// either; the user fixes their cert server-side.

public struct DavEntry: Hashable, Sendable {
    public let href: String        // URL-decoded server path
    public let name: String
    public let isDirectory: Bool
    public let size: Int64?
    public let modifiedAt: Date?
}

public enum WebDavError: LocalizedError {
    case notConfigured, http(Int, String?), transport(String), decoding(String)
    public var errorDescription: String? {
        switch self {
        case .notConfigured:        return "Nextcloud credentials not set."
        case .http(let s, _):       return "Server returned HTTP \(s)."
        case .transport(let m):     return "Network error: \(m)."
        case .decoding(let m):      return "Couldn't parse response: \(m)."
        }
    }
}

public actor WebDavClient {
    let creds: WebDavCreds
    private let session: URLSession
    public init(_ c: WebDavCreds, session: URLSession = .shared) {
        self.creds = c; self.session = session
    }

    private func userRoot(_ path: String = "") -> URL? {
        let base = creds.baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/ "))
        let cleaned = path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        var s = "\(base)/remote.php/dav/files/\(creds.username)"
        if !cleaned.isEmpty { s += "/\(cleaned)" }
        return URL(string: s)
    }

    private func basicAuth() -> String {
        let token = Data("\(creds.username):\(creds.appPassword)".utf8).base64EncodedString()
        return "Basic \(token)"
    }

    public func testConnection() async throws -> Int {
        try await list("").count
    }

    public func list(_ path: String) async throws -> [DavEntry] {
        guard let url = userRoot(path) else { throw WebDavError.notConfigured }
        var req = URLRequest(url: url)
        req.httpMethod = "PROPFIND"
        req.setValue(basicAuth(), forHTTPHeaderField: "Authorization")
        req.setValue("1", forHTTPHeaderField: "Depth")
        req.setValue("application/xml; charset=utf-8", forHTTPHeaderField: "Content-Type")
        req.httpBody = """
        <?xml version="1.0"?>
        <d:propfind xmlns:d="DAV:">
          <d:prop>
            <d:displayname/>
            <d:getcontentlength/>
            <d:getlastmodified/>
            <d:resourcetype/>
          </d:prop>
        </d:propfind>
        """.data(using: .utf8)

        let (data, resp) = try await send(req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard code == 207 || code == 200 else {
            throw WebDavError.http(code, String(data: data, encoding: .utf8))
        }
        return try parseMultiStatus(data, selfPath: url.path)
    }

    public func get(_ path: String) async throws -> Data {
        guard let url = userRoot(path) else { throw WebDavError.notConfigured }
        var req = URLRequest(url: url)
        req.setValue(basicAuth(), forHTTPHeaderField: "Authorization")
        let (data, resp) = try await send(req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard code == 200 else { throw WebDavError.http(code, String(data: data, encoding: .utf8)) }
        return data
    }

    public func put(directory dir: String, file name: String, data: Data,
                    contentType: String = "application/octet-stream") async throws {
        let cleaned = dir.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if !cleaned.isEmpty {
            let parts = cleaned.split(separator: "/")
            var acc = ""
            for p in parts {
                acc = acc.isEmpty ? String(p) : "\(acc)/\(p)"
                try await mkcol(acc)
            }
        }
        let dest = cleaned.isEmpty ? name : "\(cleaned)/\(name)"
        guard let url = userRoot(dest) else { throw WebDavError.notConfigured }
        var req = URLRequest(url: url)
        req.httpMethod = "PUT"
        req.setValue(basicAuth(), forHTTPHeaderField: "Authorization")
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")
        req.httpBody = data
        let (body, resp) = try await send(req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        if !(200..<300).contains(code) {
            throw WebDavError.http(code, String(data: body, encoding: .utf8))
        }
    }

    private func mkcol(_ path: String) async throws {
        guard let url = userRoot(path) else { throw WebDavError.notConfigured }
        var req = URLRequest(url: url)
        req.httpMethod = "MKCOL"
        req.setValue(basicAuth(), forHTTPHeaderField: "Authorization")
        let (body, resp) = try await send(req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        if code == 405 || code == 409 { return }   // already exists
        if !(200..<300).contains(code) {
            throw WebDavError.http(code, String(data: body, encoding: .utf8))
        }
    }

    private func send(_ req: URLRequest) async throws -> (Data, URLResponse) {
        do { return try await session.data(for: req) }
        catch { throw WebDavError.transport(error.localizedDescription) }
    }

    // MARK: - PROPFIND parsing
    private func parseMultiStatus(_ data: Data, selfPath: String) throws -> [DavEntry] {
        let parser = DavParser(selfPath: selfPath)
        let x = XMLParser(data: data)
        x.delegate = parser
        x.shouldProcessNamespaces = true
        guard x.parse() else {
            throw WebDavError.decoding(x.parserError?.localizedDescription ?? "?")
        }
        return parser.entries.sorted {
            if $0.isDirectory != $1.isDirectory { return $0.isDirectory }
            return $0.name.lowercased() < $1.name.lowercased()
        }
    }
}

// XMLParser delegate is reference-only, so we keep it out of the actor.
private final class DavParser: NSObject, XMLParserDelegate {
    private let httpDate: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "GMT")
        f.dateFormat = "EEE, dd MMM yyyy HH:mm:ss zzz"
        return f
    }()

    let selfPath: String
    var entries: [DavEntry] = []
    private var inResponse = false
    private var current = ""
    private var href: String?
    private var displayname: String?
    private var size: Int64?
    private var lastModified: String?
    private var isDir = false
    init(selfPath: String) { self.selfPath = selfPath }

    func parser(_ parser: XMLParser, didStartElement name: String,
                namespaceURI: String?, qualifiedName qName: String?, attributes: [String:String] = [:]) {
        switch name.lowercased() {
        case "response":
            inResponse = true
            href = nil; displayname = nil; size = nil; lastModified = nil; isDir = false
        case "collection":
            if inResponse { isDir = true }
        default: break
        }
        current = ""
    }
    func parser(_ parser: XMLParser, foundCharacters text: String) {
        current += text
    }
    func parser(_ parser: XMLParser, didEndElement name: String,
                namespaceURI: String?, qualifiedName qName: String?) {
        let trimmed = current.trimmingCharacters(in: .whitespacesAndNewlines)
        switch name.lowercased() {
        case "href":               if inResponse { href = trimmed }
        case "displayname":        if inResponse { displayname = trimmed }
        case "getcontentlength":   if inResponse { size = Int64(trimmed) }
        case "getlastmodified":    if inResponse { lastModified = trimmed }
        case "response":
            guard let raw = href else { inResponse = false; return }
            let decoded = raw.removingPercentEncoding ?? raw
            let pathOnly = URL(string: raw)?.path ?? raw
            let self0 = selfPath.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            let cand0 = pathOnly.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            if cand0 != self0 {
                let trimmedTail = decoded.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
                let displayName = (displayname?.isEmpty == false ? displayname : nil) ??
                    (trimmedTail.split(separator: "/").last.map(String.init) ?? trimmedTail)
                entries.append(DavEntry(
                    href: decoded,
                    name: displayName,
                    isDirectory: isDir,
                    size: size,
                    modifiedAt: lastModified.flatMap { httpDate.date(from: $0) }
                ))
            }
            inResponse = false
        default: break
        }
    }
}
